### ISSUE TITLE
Add source locations to annotations

### ISSUE THREAD

**Issue by smolkaj** [p4lang/p4runtime#285]
## Motivation
The type of (unstructured) annotations in `p4info.proto` is currently just `string`, e.g.:
```
message Preamble {
  uint32 id = 1;
  string name = 2;
  string alias = 3;
  repeated string annotations = 4;
  Documentation doc = 5;
}
```
Since annotations can be arbitrary strings, consuming them will typically involve parsing them first. Thus, an annotation consumer may want to point out syntax/type errors when parsing fails. For example, our [p4-constraints prototype](https://github.com/p4lang/p4-constraints) produces messages such as the following:
```
path/to/example.p4:60:5-15:
   |   @entry_restriction("
60 |     ternary_key::prefix_length == 32
   |     ^^^^^^^^^^^
Type error: expression of type ternary<32> has no field 'prefix_length'
```
This requires knowledge of the location of the annotation in the source file.

## Proposal
I propose adding a more informative `Annotation` message as follows:
```
message Annotation {
  string tag = 1;                        // The part between '@' and '('.
  string payload = 2;                    // The part between '(' and ')'.
  SourceLocation source_location = 3;
}
message SourceLocation {
  string file = 1;  // Path to source file (total or relative to working directory).
  int line = 2;     // Offset of payload in source file.
  int column = 3;   // Offset of payload in source file.
}
```
`Preamble` (and other messages carrying annotations) would then be updated as follows:
```
message Preamble {
  ...
  repeated Annotation annotations = 4;
  ...
}
```

## Workaround
Our prototype currently obtains source locations using the builtin preporcessor macros `__FILE__` and `__LINE__` as follows:
```
@file(__FILE__)
@line(__LINE__)
@my_annotation("
  foo bar baz
")
```
This pollutes the .p4 source file with clutter, is a burden to the user, and is brittle.

**Comment by stefanheule** [comment]
Shouldn't we add this to both structured and unstructured annotations?

**Comment by smolkaj** [comment]
We discussed this proposal at the P4 API working group meeting. We agreed on the following points:
* Source locations should be added to both structured and unstructured annotations.
  * While syntax errors seem unlikely in structured annotations, there will likely still exists semantically invalid inputs; being able to point out the source location of such invalid inputs seems useful.
* The source location should refer to the location of the '@' symbol in all cases.
* To make source locations more useful, [p4c needs to preserve whitespace](https://github.com/p4lang/p4c/issues/2333).
* My initial proposal breaks backwards compatibility, so it cannot be added until p4runtime 2.0.
* In the meantime, we can add source locations for unstructured annotations in a backwards compatible way as follows:  
  ```
  message Preamble {
    ...
    repeated string annotations = 4;
    Documentation doc = 5;
    repeated SourceLocation annotation_locations = 6;
  }
  ```
@stefanheule @antoninbas Please let me know if I forgot or misrepresented anything.

**Comment by antoninbas** [comment]
@smolkaj looks good to me

you can also open a new issue for merging `annotations` and `annotation_locations` once your PR is merged, and we can tag it for P4Runtime v2.0

**Comment by smolkaj** [comment]
@antoninbas: At that point we should probably just have a single message for all types of annotations? See my PR.

**Comment by smolkaj** [comment]
I opened issue #288 to track merging the annotations in 2.0

**Comment by smolkaj** [comment]
The final step here is to make p4c populate the new source location messages.

**Comment by antoninbas** [comment]
@smolkaj I'm going to close this. You should open the appropriate issue in p4lang/p4c.

**Comment by smolkaj** [comment]
Done, see https://github.com/p4lang/p4c/issues/2418.
